client/state/data-layer/wpcom-http/utils.js

```diff
@@ -110,54 +105,94 @@ const defaultOptions = {
  * @param {Function} onError called if the action meta includes error data
- * @param {Object} options configures additional dispatching behaviors
- + @param {Function} [options.fromApi] maps between API data and Calypso data
- + @param {Function} [options.onProgress] called on progress events when uploading
- * @returns {?*} please ignore return values, they are undefined
+ * @param {Function} onProgress called on progress events when uploading
+ * @param {Function} fromApi maps between API data and Calypso data
+ * @returns {Action} action or action thunk to be executed in response to HTTP event
+ */
+export const dispatchRequest = options => {
+	if ( ! options.initiator ) {
+		warn( 'initiator handler is not defined: no request will ever be issued' );
+	}
+
+	if ( ! options.onSuccess ) {
+		warn( 'onSuccess handler is not defined: response to the request is being ignored' );
+	}
+
+	if ( ! options.onError ) {
+		warn( 'onError handler is not defined: error during the request is being ignored' );
+	}
+
+	return ( store, action ) => {
+		// create the low-level action we want to dispatch
+		const requestAction = createRequestAction( options, action );
+		// dispatch the low level action (if any was created) and return the result
+		return requestAction ? store.dispatch( requestAction ) : undefined;
+	};
+};
+
+/*
+ * Converts an application-level Calypso action that's handled by the data-layer middleware
+ * into a low-level action. For example, HTTP request that's being initiated, or a response
+ * action with a `meta.dataLayer` property.
  */
-export const dispatchRequest = ( initiator, onSuccess, onError, options = {} ) => (
-	store,
-	action
-) => {
-	const { fromApi, onProgress } = { ...defaultOptions, ...options };
+function createRequestAction( options, action ) {
+	const {
+		initiator = noop,
+		onSuccess = noop,
+		onError = noop,
+		onProgress = noop,
+		fromApi = identity,
+	} = options;
 
 	const error = getError( action );
-	if ( undefined !== error ) {
-		return onError( store, action, error );
+	if ( error ) {
+		return onError( action, error );
 	}
 
 	const data = getData( action );
-	if ( undefined !== data ) {
+	if ( data ) {
 		try {
-			return onSuccess( store, action, fromApi( data ) );
+			return onSuccess( action, fromApi( data ) );
 		} catch ( err ) {
-			return onError( store, action, err );
+			return onError( action, err );
 		}
 	}
 
 	const progress = getProgress( action );
-	if ( undefined !== progress ) {
-		return onProgress( store, action, progress );
+	if ( progress ) {
+		return onProgress( action, progress );
 	}
 
-	return initiator( store, action );
-};
+	const initiatorAction = initiator( action );
+	if ( ! initiatorAction ) {
+		warn( "The `initiator` handler didn't return any action: no request will be issued" );
+	}
+
+	return Object.assign( {}, initiatorAction, {
+		onSuccess: action,
+		onError: action,
+		onProgress: action,
+	} );
```

Comment:
do you think it would be good to only assign these to network-type requests? we currently have just two: `WPCOM_HTTP_REQUEST` and another raw HTTP request.

wouldn't see why normal actions need these, especially since they are going onto the top-level of the action

Reply:
I'm not sure how to distinguish between network-type requests and others. I want to avoid enumerating some "special" action types inside `dispatchRequest` and give them special treatment.

I decided to roll back this change and start passing a second `action` parameter to the `http()` action creator again. The amount of typing it saved is not that great. `dispatchRequest` doesn't decorate the initiator action in any way and only dispatches it.